sgx-hasher/src/lib.rs

```diff
@@ -38,46 +36,185 @@ impl Hasher {
+        // For each page in the input...
+        for (i, page) in data.chunks(4096).enumerate() {
+            let off = offset + i as u64 * 4096;
 
-        // Hash for the EEXTEND instruction.
-        if let sgx_traits::Flags::MEASURE = flags {
-            for i in 0..16 {
-                let poff = i * 256;
-                let eoff = offset + poff as u64;
+            // Hash for the EADD instruction.
+            self.0.update(&EADD.to_le_bytes());
+            self.0.update(&off.to_le_bytes());
+            self.0.update(&secinfo.as_ref()[..48]);
 
-                self.0.update(&EEXTEND.to_le_bytes());
-                self.0.update(&eoff.to_le_bytes());
-                self.0.update(&[0u8; 48]);
+            // Hash for the EEXTEND instruction.
+            if flags.contains(sgx_traits::Flags::MEASURE) {
+                for (j, segment) in page.chunks(256).enumerate() {
+                    let off = off + j as u64 * 256;
 
-                self.0.update(&data[poff..][..256]);
+                    self.0.update(&EEXTEND.to_le_bytes());
+                    self.0.update(&off.to_le_bytes());
+                    self.0.update(&[0u8; 48]);
+                    self.0.update(segment);
+                }
             }
         }
     }
 
     // Produces MRENCLAVE value (SHA256 of fields specified in
     // https://github.com/enarx/enarx/wiki/SGX-Measurement)
-    pub fn finish(self) -> Result<[u8; 32]> {
-        Ok(self.0.finish())
+    pub fn finish(self) -> [u8; 32] {
+        self.0.finish()
     }
 }
 
 #[cfg(test)]
 mod test {
-    // TODO: Add tests (ex. for modulus, exponent of rsa key)
+    use super::*;
+    use sgx_types::page::{Flags as Perms, SecInfo};
+    use sgx_types::{attr, misc, sig};
+
+    // A NOTE ABOUT THIS TESTING METHODOLOGY
+    //
+    // The ANSWER values in this test are not guaranteed to be correct.
+    // They were produced by getting the hasher to output the same hash
+    // for the binary from the Linux kernel SGX selftest and then running
+    // the algorithm again to produce these ANSWERs. At least some of the
+    // ANSWERs are correct. The remaining ones alert us to algorithmic
+    // changes. We expect that over time our ANSWERs will be correct.
+    //
+    // The canonical source of correctness for this algorithm is, of
+    // course, the Intel SGX CPU/microcode/ME. If you can demonstrate a
+    // a case where we don't match this, we will happily change our ANSWERs.
+
+    const DATA: [u8; 4096] = [123u8; 4096];
+    const LONG: [u8; 8192] = [123u8; 8192];
+
+    fn hash(input: &[(&[u8], SecInfo)]) -> [u8; 32] {
+        let size = input.iter().fold(0, |c, x| c + x.0.len() as u64);
+
+        let contents = sig::Contents::new(
+            misc::MiscSelect::default(),
+            attr::Attributes::default(),
+            [0u8; 32],
+            0,
+            0,
+        );
+
+        let secs = secs::Secs::new(0, size.next_power_of_two(), 1, &contents);
```

Comment:
So the `Secs` has a base of `0`, a size that is larger than the input size by a power of two, an `ssa` size of `1`, and the default `contents` from above. It may be helpful to include this as a comment.

Reply:
I added a comment in the latest push.